### PR TITLE
Correctly deal with unicode characters

### PR DESCRIPTION
**Pull request by jviotti**
This seems to be the cause of many Etcher bugs. This commit deals with
the following issues:

- Unicode characters
- Non printable ascii characters

Signed-off-by: Juan Cruz Viotti <jviotti@openmailbox.org>

### CODE REVIEW

File: tests/parse.spec.js [hunk 1]
```diff
@@ -232,4 +232,54 @@ describe('Parse', function() {
...
+      'description: "foo\bbar\rbaz"'
+    ].join('\n'))).to.deep.equal([
+      {
+        device: '/dev/disk1',
+        description: 'foobarbaz'
+      }
+    ]);
+  });
+
+  it('should not remove tabs inside values', function() {
```
Comment (lurch) [inline tests/parse.spec.js]: Huh? Seems like it's actually replacing tabs by 4 spaces. If that _is_ the intended behaviour, this test should probably be renamed ;-)

File: tests/parse.spec.js [hunk 2]
```diff
@@ -232,4 +232,54 @@ describe('Parse', function() {
...
+      'description: "\\]=-01`234567890=-098e"'
+    ].join('\n'))).to.deep.equal([
+      {
+        device: '/dev/disk1',
+        description: ']=-01`234567890=-098e'
+      }
+    ]);
+  });
+
+  it('should delete known escape sequences', function() {
```
Comment (lurch) [inline tests/parse.spec.js]: Does this impact on the changes I made in #156 ?

Comment (lurch) [inline tests/parse.spec.js]: (and if TAB is a special-case, maybe this test should be called `it('should delete known escape sequences except TAB', function() {` ?)

File: lib/parse.js [hunk 3]
```diff
@@ -65,23 +65,36 @@ module.exports = (input) => {
...
+            if (string[index - 1] === '\\') {
+              return match;
+            }
+
+            return _.last(match);
+          })
+
+          // Remove non printable ascii characters
+          // See http://stackoverflow.com/a/24229554
+          .replace(/[^\x20-\x7E]+/g, '')
```
Comment (lurch) [inline lib/parse.js]: Heh, very similar to the regex at https://github.com/resin-io/etcher/blob/master/scripts/ci/ensure-all-text-files-only-ascii.sh#L31 ;-)

File: tests/parse.spec.js [hunk 4]
```diff
@@ -232,4 +232,54 @@ describe('Parse', function() {
     ]);
   });
 
+  it('should handle unknown escape sequences inside a value', function() {
```
Comment (lurch) [inline tests/parse.spec.js]: IMHO we should also have a test on a string containing multiple unknown escape sequences

Comment (lurch) [inline tests/parse.spec.js]: ...and what exactly does "handle" mean here?

File: tests/parse.spec.js [hunk 5]
```diff
@@ -232,4 +232,54 @@ describe('Parse', function() {
...
+      'description: "foo\tbar"'
+    ].join('\n'))).to.deep.equal([
+      {
+        device: '/dev/disk1',
+        description: 'foo    bar'
+      }
+    ]);
+  });
+
+  it('should handle unicode characters inside a value', function() {
```
Comment (lurch) [inline tests/parse.spec.js]: Should this also be testing printable unicode characters in addition to unprintable unicode characters?
And does "should handle" here actually mean "should delete"?

**Comment by lurch** [comment]
Do we have any idea where these unicode characters and unknown escape sequences are actually coming from?

File: tests/parse.spec.js (hunk 4 above)
Comment (jviotti) [inline tests/parse.spec.js]: I spammed the string with more backslashes, and clarified what "handle" means!

File: tests/parse.spec.js (hunk 2 above)
Comment (jviotti) [inline tests/parse.spec.js]: I think the changes are complementary. Your changes make the linux script converts things like `\011` into ASCII escape sequences that can be removed here, but do you think we can move that sed logic here, so we make sure it gets applied everywhere?

> (and if TAB is a special-case, maybe this test should be called it('should delete known escape sequences except TAB', function() { ?)

Fixed!

File: tests/parse.spec.js (hunk 1 above)
Comment (jviotti) [inline tests/parse.spec.js]: Yeah, good catch. Fixed!

File: tests/parse.spec.js (hunk 5 above)
Comment (jviotti) [inline tests/parse.spec.js]: Makes sense. I added some printable characters, and clarified the message.

**Comment by jviotti** [comment]
@lurch All the drive descriptions I added to the test suite are legit descriptions I obtained from Sentry. Interestingly enough, the 92% of these occurrences come from GNU/Linux systems, but I have no clue how they got there in the first place.

File: tests/parse.spec.js (hunk 2 above)
Comment (lurch) [inline tests/parse.spec.js]: Yeah, that's a good point. My changes in #156 convert the Linux mount-specific escape sequences to 'regular ASCII escapes', and then your changes here remove those regular ASCII escape sequences. That sounds like the correct separation of logic.

Although now I think about it, perhaps the changes you're making in this PR are Etcher-specific, and there might be other users of `drivelist` who _want_ these escape sequences to be preserved? Perhaps there should be an optional parameter that a user could supply to drivelist, specifying whether they want these escape sequences preserved or not?

**Comment by lurch** [comment]
Hmmm, interesting. I wonder if any of them might be caused by data-corruption, and perhaps Linux is more tolerant at 'preserving' corrupted data and passing it back to the user? Or maybe it's Windows or OSX specific codepage issues which Linux isn't able to properly handle? Who knows... ;-) \*shrug\*

File: tests/parse.spec.js (hunk 2 above)
Comment (jviotti) [inline tests/parse.spec.js]: > Perhaps there should be an optional parameter that a user could supply to drivelist, specifying whether they want these escape sequences preserved or not?

I think it makes a lot of sense. I made the escape sequences go through without issues, and we can handle from the Etcher side if really necessary.
